packetbeat/protos/dns/dns.go

```diff
@@ -383,240 +392,229 @@ func (dns *Dns) expireTransaction(t *DnsTransaction) {
-			mapStr["preference"] = r.MX.Preference
-			mapStr["data"] = nameToString(r.MX.Name)
-		case layers.DNSTypeSRV:
-			mapStr["priority"] = r.SRV.Priority
-			mapStr["weight"] = r.SRV.Weight
-			mapStr["port"] = r.SRV.Port
-			mapStr["data"] = nameToString(r.SRV.Name)
-		case layers.DNSTypeCNAME:
-			mapStr["data"] = nameToString(r.CNAME)
-		case layers.DNSTypePTR:
-			mapStr["data"] = nameToString(r.PTR)
-		case layers.DNSTypeNS:
-			mapStr["data"] = nameToString(r.NS)
+			// We don't have special handling for this type
+			logp.Debug("dns", "Unsupported RR type %s", dnsTypeToString(rrType))
+		case *mkdns.A:
+			mapStr["data"] = x.A.String()
+		case *mkdns.AAAA:
+			mapStr["data"] = x.AAAA.String()
+		case *mkdns.CNAME:
+			mapStr["data"] = x.Target
+		case *mkdns.MX:
+			mapStr["preference"] = x.Preference
+			mapStr["data"] = x.Mx
+		case *mkdns.NS:
+			mapStr["data"] = x.Ns
+		case *mkdns.PTR:
+			mapStr["data"] = x.Ptr
+		case *mkdns.RFC3597:
+			// Miekg/dns lib doesn't handle this type
+			logp.Debug("dns", "Unknown RR type %s", dnsTypeToString(rrType))
+			mapStr["data"] = x.Rdata
+		case *mkdns.SOA:
+			mapStr["rname"] = x.Mbox
+			mapStr["serial"] = x.Serial
+			mapStr["refresh"] = x.Refresh
+			mapStr["retry"] = x.Retry
+			mapStr["expire"] = x.Expire
+			mapStr["minttl"] = x.Minttl
+			mapStr["data"] = x.Ns
+		case *mkdns.SRV:
+			mapStr["priority"] = x.Priority
+			mapStr["weight"] = x.Weight
+			mapStr["port"] = x.Port
+			mapStr["data"] = x.Target
+		case *mkdns.TXT:
+			mapStr["data"] = strings.Join(x.Txt, " ")
 		}
 	}
 
 	return mapStrArray
 }
 
-// dnsQuestionToString converts a DNSQuestion to a string.
-func dnsQuestionToString(q layers.DNSQuestion) string {
-	name := nameToString(q.Name)
-	if len(name) == 0 {
-		name = "Root"
-	}
-	return fmt.Sprintf("class %s, type %s, %s", dnsClassToString(q.Class),
-		dnsTypeToString(q.Type), name)
+// dnsQuestionToString converts a Question to a string.
+func dnsQuestionToString(q mkdns.Question) string {
+	name := q.Name
+
+	return fmt.Sprintf("class %s, type %s, %s", dnsClassToString(q.Qclass),
+		dnsTypeToString(q.Qtype), name)
 }
 
-// dnsResourceRecordToString converts a DNSResourceRecord to a string.
-func dnsResourceRecordToString(rr *layers.DNSResourceRecord) string {
-	name := nameToString(rr.Name)
-	if len(name) == 0 {
-		name = "Root"
-	}
+// dnsResourceRecordToString converts a RR to a string.
+func dnsResourceRecordToString(rr mkdns.RR) string {
+	rrHeader := rr.Header()
+	rrType := rrHeader.Rrtype
+
 	var data string
-	switch rr.Type {
+	switch x := rr.(type) {
 	default:
-		// We don't have special handling for this type so use the same
-		// encoding used for names to output the raw data for this type.
-		data = nameToString(rr.Data)
-	case layers.DNSTypeA, layers.DNSTypeAAAA:
-		data = rr.IP.String()
-	case layers.DNSTypeSOA:
+		// We don't have special handling for this type
+		logp.Debug("dns", "Unsupported RR type %s", dnsTypeToString(rrType))
```

Comment:
Yes. At first I thought that couldn't be achieved ... because only [type RFC3597](https://github.com/miekg/dns/blob/master/types.go#L1044-L1048) has the Rdata field, other types having the fields according to their format.
But a method [ToRFC3597](https://github.com/miekg/dns/blob/master/dns.go#L85-#L100) helps fetch this field from any RR type. 
With some lines of code ([e.g. here](https://github.com/elastic/beats/blob/551ee4b13d39f8c779e2b6da12fcb966ee27db03/packetbeat/protos/dns/dns.go#L452-L463)), the Rdata is fetched. I'm not sure a simpler exists though.